api/docs/scatter_gather_emulation.dox

```diff
@@ -0,0 +1,230 @@
+
+- When a register is written by an application instruction, we invalidate all `spill_slot_to_reg` entries that are mapped to that register, and also set `gpr_is_native` for that register.
+- When a register is written by a non-drreg meta instruction, we clear `gpr_is_native` for that reg.
+- When a register is loaded by drreg from the slot it was spilled to, we set `gpr_is_native`.
+- When a register is spilled to some spill slot, we set `spill_slot_to_reg` for that spill slot to that reg.
+
+This strategy allows us to robustly keep track of the various corner cases that can arise in drreg, like spill regions from different phases overlapping (nesting or just overlapping), and the other known issues linked above. This was implemented by this [commit](https://github.com/DynamoRIO/dynamorio/commit/f62441bc2a0ced41263e2229deb3691433d6abb9).
+
+The drawback of this approach is that it needs to be aware of other methods of spilling and restoring registers outside drreg ([dropped PR](https://github.com/DynamoRIO/dynamorio/pull/4987)). DynamoRIO uses various such methods internally (spilling to stack, slots not managed by drreg), and also the client may use their own unique methods. So, some non-drreg meta instructions may actually restore an application value to a register, but this approach will not be able to recognize that. This may cause it to lose track of some register’s application value. We dropped this approach on encountering [DynamoRIO/dynamorio#4963](https://github.com/DynamoRIO/dynamorio/issues/4963).
+
+#### Preferred: Pairing restores with spills (instead of the other way)
+
+The key observation behind this approach is that it is easier to find the matching spill for a given restore, than to find the matching restore for a given spill. This is because there may be other restores besides the final restore, e.g. restores for app read, user prompted restores, etc. This makes it hard to find exactly where the spill region for a register/aflags ends.  Additional complexities include the fact that aflags re-spills may not use the same slot, which makes differentiating spills from multiple phases difficult.
+
+Each restore must have a matching spill. Based on this observation, we scan the faulting fragment from end to beginning, matching register restores to their spills. When we reach the faulting instruction, any restore for which we did not see the matching spill yet must be performed by the drreg state restoration. This was implemented by ([commit](https://github.com/DynamoRIO/dynamorio/commit/be78c124eb787601182e3c73f4be8bb859c50ef8)).
```

Comment:
What about a register with two spills (the 2nd is after an app write) and one restore, with the fault between the 2 spills?  Won't this approach incorrectly fail to restore the register?

Reply:
That's right; the approach will not restore that reg. However, as the reg is dead before the app write, it does not affect app correctness, unless the app deliberately inspects that reg I guess.

Reply:
If drreg_options_t.conservative is turned on, drreg will then do the wrong thing here.  This should probably be documented in drreg.h, and mentioned here.  (Not considered likely to affect most apps of course.)

Reply:
This seems very hard to do without additional metadata. Added a comment to #3801, changed the doc and added the comment to drreg.h too. Sad to see an unhandled case (even though less likely to matter) in drreg state restoration. I'm finding it very tricky to get all corner cases working perfectly without additional  metadata.